File: python/tests/test_polyfuncs.py
import unittest

from ring import BoolPolyRing
from polyfuncs import spoly, normal_form


class TestPolyUtils(object):

    def test_spoly(self):
        a, b, c, d = self.B.gens

        _1 = self.B.one

        s = spoly(a*b*c, a*b + _1)
        self.assertEqual(s, c)

        s = spoly(a*b*c + _1, a*b + _1)
        self.assertEqual(s, c + _1)

        f = a*b*c + c*d + a*b + _1
        g = c*d + b

        s = spoly(f, g)
        self.assertEqual(s, a*b + a*c*d + c*d + _1)

    def test_normal_form_1(self):
        x0, x1, x2, x3 = self.B.gens

        _1 = self.B.one

        F = [
            x0 + x1 + x2 + x3,
            x0*x1 + x1*x2 + x0*x3 + x2*x3,
            x0*x1*x2 + x0*x1*x3 + x0*x2*x3 + x1*x2*x3,
            x0*x1*x2*x3 + _1
        ]

        p = x0*x1*x2 + x0*x1*x3 + x0*x2*x3 + x2

        p_nf = normal_form(p, F)

        self.assertEqual(p_nf, x1*x2*x3 + x2)

    def test_normal_form_2(self):
        x0, x1, x2, x3 = self.B.gens

        _1 = self.B.one
        _0 = self.B.zero
        # Groebner basis for F should reduce to 0
        G = [x0 + _1, x1 + _1, x2 + _1, x3 + _1]

        p = x0*x1*x2 + x0*x1*x3 + x0*x2*x3 + x2

        p_nf = normal_form(p, G)

        self.assertEqual(_0, p_nf)

    def test_normal_form_3(self):
        x0, x1, x2, x3 = self.B.gens

        _1 = self.B.one

        F = [
            x0 + x1 + x2 + x3,
            x0*x1 + x0*x3 + x1*x2 + x2*x3,
            x0*x1*x2 + x0*x1*x3 + x0*x2*x3 + x1*x2*x3,
            x0*x1*x2*x3 + _1,
            x1 + x3
        ]

        p = x1*x2*x3 + _1

        p_nf = normal_form(p, F)

        self.assertEqual(p_nf, x2*x3 + _1)

    def test_normal_form_4(self):
        x1, x2, x3 = self.B.gens[:3]

        _1 = self.B.one

        F = [
            x1*x2*x3,
            x1*x2*x3 + x1*x2,
            x1*x2*x3 + x1*x3,
            x1*x2*x3 + x1*x2 + x1*x3 + x1,
            x1*x2*x3 + x2*x3,
            x1*x2*x3 + x1*x2 + x2*x3 + x2,
            x1*x2*x3 + x1*x3 + x2*x3 + x3,
            x1*x2*x3 + x1*x2 + x1*x3 + x1 + x2*x3 + x2 + x3 + _1
        ]

        # S-polynomial for p = x1*x2*x3 + x1*x2 and
        # q = x1*x2*x3 + x1*x2 + x1*x3 + x1 + x2*x3 + x2 + x3 + 1
        s = x1*x3 + x1 + x2*x3 + x2 + x3 + _1

        nf = normal_form(s, F)

        self.assertEqual(s, nf)

    def test_normal_form_5(self):
        x1, x2, x3 = self.B.gens[:3]

        _1 = self.B.one

        s = x1*x3 + x1 + x2*x3 + x2 + x3 + _1

        F = [
            x1*x2*x3,
            x1*x2*x3 + x1*x2,
            x1*x2*x3 + x1*x3,
            x1*x2*x3 + x1*x2 + x1*x3 + x1,
            x1*x2*x3 + x2*x3,
            x1*x2*x3 + x1*x2 + x2*x3 + x2,
            x1*x2*x3 + x1*x3 + x2*x3 + x3,
            x1*x2*x3 + x1*x2 + x1*x3 + x1 + x2*x3 + x2 + x3 + _1,
            x1*x3,
            x2*x3,
            x3,
            x1*x2,
            x2,
            x1,
            _1
        ]

        # Should reduce to 0
        nf = normal_form(s, F)

        self.assertTrue(nf.isZero())


class TestPolyListUtils(unittest.TestCase, TestPolyUtils):

    def setUp(self):
        self.B = BoolPolyRing(4, poly_type="list")


class TestPolyZDDUtils(unittest.TestCase, TestPolyUtils):

    def setUp(self):
        self.B = BoolPolyRing(4, poly_type="zdd")

File: symbsat-cpp/tests/test_polyfuncs.cpp
#include "catch.hpp"

#include "monom.h"
#include "poly.h"
#include "zdd.h"
#include "polyfuncs.h"

using namespace symbsat;

TEST_CASE("S-polynomial PolyZDD", "[spoly-polyzdd]") {
  ZDD<Monom32> f, g;
  ZDD<Monom32> a(1), b(2), c(3), d(4);
  ZDD<Monom32> s, s1, s2;
  ZDD<Monom32> _1; _1.setOne();

  s = spoly<ZDD<Monom32>>(f, g);
  REQUIRE(s.isZero());

  s1 = spoly<ZDD<Monom32>>(a*b*c, a*b + _1);
  REQUIRE(s1 == c);

  s = spoly<ZDD<Monom32>>(a*b*c + _1, a*b + _1);

  REQUIRE(s == c + _1);

  ZDD<Monom32> f1 = a*b*c + c*d + a*b + _1;
  ZDD<Monom32> g1 = c*d + b;

  s = spoly(f1, g1);
  REQUIRE(s == a*b + a*c*d + c*d + _1);
}

TEST_CASE("S-polynomial PolyList", "[spoly-polylist]") {
  Poly<Monom32> f, g;
  Poly<Monom32> a(1), b(2), c(3), d(4);
  Poly<Monom32> s, s1, s2, s3;
  Poly<Monom32> _1; _1.setOne();

  s = spoly<Poly<Monom32>>(f, g);
  REQUIRE(s.isZero());

  s1 = spoly<Poly<Monom32>>(a*b*c, a*b + _1);
  REQUIRE(s1 == c);

  s2 = spoly<Poly<Monom32>>(a*b*c + _1, a*b + _1);

  REQUIRE(s == c + _1);

  Poly<Monom32> f1 = a*b*c + c*d + a*b + _1;
  Poly<Monom32> g1 = c*d + b;

  s3 = spoly(f1, g1);
  REQUIRE(s == a*b + a*c*d + c*d + _1);
}

TEST_CASE("NormalForm 1 PolyList", "[normalform-polylist]") {
  Poly<Monom32> x1(1), x2(2),
                x3(3), x4(4), _1;
  _1.setOne();

  Poly<Monom32> p = x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x3;
  Poly<Monom32> p_nf;

  std::vector<Poly<Monom32>> F {
      x1 + x2 + x3 + x4,
      x1*x2 + x2*x3 + x1*x3 + x3*x4,
      x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x2*x3*x4,
      x1*x2*x3*x4 + _1
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf == x2*x3*x4 + x3);
}

TEST_CASE("NormalForm 1 PolyZDD", "[normalform-polyzdd]") {
  ZDD<Monom32> x1(1), x2(2),
               x3(3), x4(4), _1;
  _1.setOne();

  ZDD<Monom32> p = x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x3;
  ZDD<Monom32> p_nf;

  std::vector<ZDD<Monom32>> F {
      x1 + x2 + x3 + x4,
      x1*x2 + x2*x3 + x1*x3 + x3*x4,
      x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x2*x3*x4,
      x1*x2*x3*x4 + _1
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf == x2*x3*x4 + x3);
}

TEST_CASE("NormalForm 2 PolyList", "[normalform2-polylist]") {
  Poly<Monom32> x1(1), x2(2),
                x3(3), x4(4), _1;
  _1.setOne();

  Poly<Monom32> p = x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x3;
  Poly<Monom32> p_nf;

  std::vector<Poly<Monom32>> F {
    x1 + _1,
    x2 + _1,
    x3 + _1,
    x4 + _1
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf.isZero());
}

TEST_CASE("NormalForm 2 PolyZDD", "[normalform2-polyzdd]") {
  ZDD<Monom32> x1(1), x2(2),
               x3(3), x4(4), _1;
  _1.setOne();

  ZDD<Monom32> p = x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x3;
  ZDD<Monom32> p_nf;

  std::vector<ZDD<Monom32>> F {
    x1 + _1,
    x2 + _1,
    x3 + _1,
    x4 + _1
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf.isZero());
}

TEST_CASE("NormalForm 3 PolyList", "[normalform3-polylist]") {
  Poly<Monom32> x1(1), x2(2),
                x3(3), x4(4), _1;
  _1.setOne();

  Poly<Monom32> p = x2*x3*x4 + _1;
  Poly<Monom32> p_nf;

  std::vector<Poly<Monom32>> F {
    x1 + x2 + x3 + x4,
    x1*x2 + x2*x3 + x1*x3 + x3*x4,
    x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x2*x3*x4,
    x1*x2*x3*x4 + _1,
    x2 + x4
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf == x3*x4 + _1);
}

TEST_CASE("NormalForm 3 PolyZDD", "[normalform3-polyzdd]") {
  ZDD<Monom32> x1(1), x2(2),
               x3(3), x4(4), _1;
  _1.setOne();

  ZDD<Monom32> p = x2*x3*x4 + _1;
  ZDD<Monom32> p_nf;

  std::vector<ZDD<Monom32>> F {
    x1 + x2 + x3 + x4,
    x1*x2 + x2*x3 + x1*x3 + x3*x4,
    x1*x2*x3 + x1*x2*x4 + x1*x3*x4 + x2*x3*x4,
    x1*x2*x3*x4 + _1,
    x2 + x4
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf == x3*x4 + _1);
}

TEST_CASE("NormalForm 4 PolyZDD", "[normalform4-polyzdd]") {
  ZDD<Monom32> x1(1), x2(2),
               x3(3), x4(4), _1;
  _1.setOne();

  ZDD<Monom32> p = x1 + x2;
  ZDD<Monom32> p_nf;

  std::vector<ZDD<Monom32>> F {
    x2 + x4,
    x4 + _1
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf == x1 + _1);
}

TEST_CASE("NormalForm 4 PolyList", "[normalform4-polylist]") {
  Poly<Monom32> x1(1), x2(2),
                x3(3), x4(4), _1;
  _1.setOne();

  Poly<Monom32> p = x1 + x2;
  Poly<Monom32> p_nf;

  std::vector<Poly<Monom32>> F {
    x2 + x4,
    x4 + _1
  };

  p_nf = normalform(p, F);

  REQUIRE(p_nf == x1 + _1);
}
